java-manta-client-unshaded/src/main/java/com/joyent/manta/client/crypto/AbstractAesCipherDetails.java

```diff
@@ -163,30 +163,47 @@ public String getCipherAlgorithm() {
 
     @Override
     public Cipher getCipher() {
-        if (ExternalSecurityProviderLoader.getPkcs11Provider() == null) {
-            return SupportedCipherDetails.findCipher(cipherAlgorithmJavaName,
-                    ExternalSecurityProviderLoader.getBouncyCastleProvider());
-        }
-
-        final Provider provider;
-
-        if (ExternalSecurityProviderLoader.getPkcs11Provider().containsKey(
-                "Cipher." + cipherAlgorithmJavaName)) {
-            provider = ExternalSecurityProviderLoader.getPkcs11Provider();
+        final Provider preferredProvider = ExternalSecurityProviderLoader.getPreferredProvider();
+        final Provider pkcs11Provider = ExternalSecurityProviderLoader.getPkcs11Provider();
+        final Cipher cipher;
+
+        if (pkcs11Provider != null && preferredProvider == pkcs11Provider) {
+            /* Unfortunately, libnss via pkcs11 uses different strings to
```

Comment:
If the string names don't match, how did NSS work at all before?

Reply:
The tricky bit is that the strings don't match **only** when doing `containsKey` and not when doing  `Cipher.getInstance(cipherName, provider)`.